bundle/src/main/java/org/bf2/kas/fleetshard/bundle/BundleAssembler.java

```diff
@@ -0,0 +1,526 @@
+    BundleSource syncBundle;
+    List<Map<String, ?>> almExamples;
+
+    public BundleAssembler(Config config) {
+        this.config = config;
+        version = config.getOptionalValue(Configs.BUNDLE_VERSION, String.class).orElse("999.999.999").toLowerCase();
+    }
+
+    void initialize() throws IOException {
+        bundleDir = Files.createDirectories(Path.of(config.getValue(Configs.BUNDLE_OUTPUT_DIR, String.class)));
+        Files.createDirectories(bundleDir.resolve("manifests"));
+        Files.createDirectories(bundleDir.resolve(METADATA));
+        operatorBundle = loadSource(new File(config.getValue(Configs.BUNDLE_OPERATOR_ARCHIVE, String.class)));
+        syncBundle = loadSource(new File(config.getValue(Configs.BUNDLE_SYNC_ARCHIVE, String.class)));
+    }
+
+    BundleSource loadSource(File sourceFile) throws IOException {
+        if (sourceFile.isDirectory()) {
+            return new DirectorySource(sourceFile);
+        }
+        return new ZipFileSource(new ZipFile(sourceFile));
+    }
+
+    void complete() throws IOException {
+        if (operatorBundle != null) {
+            operatorBundle.close();
+        }
+        if (syncBundle != null) {
+            syncBundle.close();
+        }
+    }
+
+    void copyResources() throws IOException {
+        copyCustomResourceDefinitions();
+        copyPriorityClasses();
+        write(bundleDir.resolve(ANNOTATIONS_PATH), operatorBundle.readString(ANNOTATIONS_PATH));
+    }
+
+    void copyCustomResourceDefinitions() throws IOException {
+        almExamples = operatorBundle.entries()
+            .filter(entryName -> entryName.endsWith(".crd.yml"))
+            .flatMap(entryName -> {
+                String crdData = operatorBundle.readString(entryName);
+                write(bundleDir.resolve(entryName), crdData);
+
+                CustomResourceDefinition crd = Serialization.unmarshal(crdData);
+
+                return crd.getSpec().getVersions().stream().map(v ->
+                    // Use a TreeMap to ensure consistent key order for testing/comparison
+                    new TreeMap<>(Map.of(
+                            "kind", crd.getSpec().getNames().getKind(),
+                            "apiVersion", String.format("%s/%s", crd.getSpec().getGroup(), v.getName()),
+                            METADATA, Map.of("name", crd.getMetadata().getName()),
+                            "spec", Collections.emptyMap())));
+            })
+            // Sorted to ensure CRDs are in the same order for testing/comparison
+            .sorted(Comparator.<Map<?, ?>, String>comparing(crdExample -> crdExample.get("kind").toString())
+                    .thenComparing(crdExample -> crdExample.get("apiVersion").toString()))
+            .collect(Collectors.toList());
+    }
+
+    void copyPriorityClasses() throws IOException {
+        try (KubernetesClient client = new DefaultKubernetesClient()) {
+            client.load(operatorBundle.getEntry("kubernetes.yml")).get().stream()
```

Comment:
Just a heads up, later versions of fabric8 require using .items() instead of get() for this.